# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: ipc/glue/MessageChannel.cpp
# Commit: b09ae4c3a94b
# Full Hash: b09ae4c3a94b663279189c50172c1f403d499183
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2021-06-22 21:29:07
# Regressor Bug: 1706374
# File Overlap Count: 1
# Description:
#   Bug 1706374 - Part 11: Add NodeController component bridging IPC and Ports, r=handyman
#   
#   The NodeController and NodeChannel types act as the backbone connecting the
#   existing IPC logic and driving the ports routing code. Individual NodeChannel
#   objects wrap and respond to messages from IPC::Channel, and the NodeController
# ==============================================================================

diff -r 04422175004b -r b09ae4c3a94b ipc/glue/MessageChannel.cpp
--- a/ipc/glue/MessageChannel.cpp	Mon Jun 21 21:53:10 2021 +0000
+++ b/ipc/glue/MessageChannel.cpp	Mon Jun 21 21:53:10 2021 +0000
@@ -25,6 +25,7 @@
 #include "mozilla/TimeStamp.h"
 #include "mozilla/UniquePtrExtensions.h"
 #include "mozilla/dom/ScriptSettings.h"
+#include "mozilla/ipc/NodeController.h"
 #include "mozilla/ipc/ProcessChild.h"
 #include "mozilla/ipc/ProtocolUtils.h"
 #include "nsAppRunner.h"
@@ -784,6 +785,20 @@
   }
 }
 
+bool MessageChannel::Open(ScopedPort aPort, Side aSide,
+                          nsISerialEventTarget* aEventTarget) {
+  MOZ_ASSERT(!mLink, "Open() called > once");
+
+  mMonitor = new RefCountedMonitor();
+  mWorkerThread = aEventTarget ? aEventTarget : GetCurrentSerialEventTarget();
+  MOZ_ASSERT(mWorkerThread, "We should always be on a nsISerialEventTarget");
+  mListener->OnIPCChannelOpened();
+
+  mLink = MakeUnique<PortLink>(this, std::move(aPort));
+  mSide = aSide;
+  return true;
+}
+
 bool MessageChannel::Open(mozilla::UniquePtr<Transport> aTransport,
                           MessageLoop* aIOLoop, Side aSide) {
   MOZ_ASSERT(!mLink, "Open() called > once");
@@ -2629,7 +2644,6 @@
       MakeUnique<Message>(MSG_ROUTING_NONE, IMPENDING_SHUTDOWN_MESSAGE_TYPE);
   MonitorAutoLock lock(*mMonitor);
   if (Connected()) {
-    MOZ_DIAGNOSTIC_ASSERT(mIsCrossProcess);
     mLink->SendMessage(std::move(msg));
   }
 }
@@ -2917,5 +2931,23 @@
   }
 }
 
+bool MessageChannel::IsCrossProcess() const {
+  mMonitor->AssertCurrentThreadOwns();
+  return mIsCrossProcess;
+}
+
+void MessageChannel::SetIsCrossProcess(bool aIsCrossProcess) {
+  mMonitor->AssertCurrentThreadOwns();
+  if (aIsCrossProcess == mIsCrossProcess) {
+    return;
+  }
+  mIsCrossProcess = aIsCrossProcess;
+  if (mIsCrossProcess) {
+    ChannelCountReporter::Increment(mName);
+  } else {
+    ChannelCountReporter::Decrement(mName);
+  }
+}
+
 }  // namespace ipc
 }  // namespace mozilla